src/main/java/com/mraof/minestuck/computer/editmode/EditToolDrag.java

```diff
@@ -0,0 +1,356 @@
+					
+					cap.setEditPos1(player.level.getBlockState(blockHit.getBlockPos()).getMaterial().isReplaceable() ? blockHit.getBlockPos() : blockHit.getBlockPos().offset(blockHit.getDirection().getNormal()));
+					cap.setEditTrace(blockHit.getLocation(), blockHit.getDirection());
+					cap.setEditReachDistance(Math.sqrt(cap.getEditPos1().distToLowCornerSqr(player.getEyePosition().x, player.getEyePosition().y, player.getEyePosition().z)));
+				}
+			}
+			if (cap.getEditPos1() != null)
+			{
+				cap.setEditPos2(getSelectionEndPoint(player, cap));
+				MSPacketHandler.sendToServer(new EditmodeFillPacket(true, isDown, cap.getEditPos1(), cap.getEditPos2(), cap.getEditTraceHit(), cap.getEditTraceDirection()));
+			}
+		}
+		else if (isDragging)
+		{
+			if (cap.getEditPos1() != null)
+			{
+				MSPacketHandler.sendToServer(new EditmodeFillPacket(true, isDown, cap.getEditPos1(), cap.getEditPos2(), cap.getEditTraceHit(), cap.getEditTraceDirection()));
+			}
+			
+			cap.resetDragTools();
+		}
+		
+		cap.setEditDragging(isDown);
+	}
+	
+	/**
+	 * Determines whether the player can use the revise tool when right-clicking, based on the block that they are holding.
+	 * @param player The client-side player.
+	 * @return True if you are in editmode and holding a non-deployable block, else false.
+	 */
+	public static boolean canEditRevise(Player player)
+	{
+		return (ClientEditHandler.isActive()
+				&& !isBlockDeployable(player)
+				&& (player.getMainHandItem().getItem() instanceof BlockItem) || (player.getOffhandItem().getItem() instanceof BlockItem));
+	}
+	
+	/**
+	 * Handles code for the recycle tool on the client-side.
+	 */
+	public static void doRecycleCode(TickEvent.ClientTickEvent event)
+	{
+		Minecraft mc = Minecraft.getInstance();
+		if (mc.player == null || event.phase == TickEvent.Phase.END)
+			return;
+		
+		Player player = mc.player;
+		IEditTools cap = player.getCapability(MSCapabilities.EDIT_TOOLS_CAPABILITY, null).orElse(null);
+		if(cap == null)
+			throw new NullPointerException("EditTool Capability is null on player " + player.getDisplayName().toString() + " on client-side!");
+		
+		if (cap.getToolMode() != null && cap.getToolMode() != IEditTools.ToolMode.RECYCLE)
+			return;
+		
+		boolean isDragging = cap.isEditDragging();
+		boolean isDown = mc.options.keyAttack.isDown();
+		
+		if (isDown)
+		{
+			if(!canEditRecycle(player))
+			{
+				cap.resetDragTools();
+				return;
+			}
+			
+			if (!isDragging)
+			{
+				BlockHitResult blockHit = getPlayerPOVHitResult(player.getLevel(), player);
+				if (blockHit.getType() == BlockHitResult.Type.BLOCK)
+				{
+					cap.setToolMode(IEditTools.ToolMode.RECYCLE);
+					
+					cap.setEditPos1(blockHit.getBlockPos());
+					cap.setEditTrace(blockHit.getLocation(), blockHit.getDirection());
+					cap.setEditReachDistance(Math.sqrt(cap.getEditPos1().distToLowCornerSqr(player.getEyePosition().x, player.getEyePosition().y, player.getEyePosition().z)));
```

Comment:
I see potential for simplifying this with a `cap.beginToolDrag(ToolMode, BlockHitResult, Player)`.

Reply:
There's also potential to creating `updateToolDrag()` and `endToolDrag()` somewhere.